src/mscorlib/src/System/Threading/ThreadPool.cs

```diff
@@ -476,57 +513,66 @@ internal bool LocalFindAndPop(IThreadPoolWorkItem callback)
+                            // Only decrement if the value is unchanged since we started looking for work
+                            // This prevents multiple threads decrementing based on overlapping scans.
+                            //
+                            // When we decrement from active, the producer may have inserted a queue item during our scan
+                            // therefore we cannot transition to empty
+                            //
+                            // When we decrement from Maybe Inactive, if the producer inserted a queue item during our scan,
+                            // the producer must write Active.  We may transition to empty briefly if we beat the
+                            // producer's write, but the producer will then overwrite us before waking threads.
+                            // So effectively we cannot mark the queue empty when an item is in the queue.
+                            Interlocked.CompareExchange(ref WorkStealingQueueList.wsqActive, wsqActiveObserved - 1, wsqActiveObserved);
                         }
-                        c--;
-                    }
-                    if ((callback == null) && !missedSteal)
-                    {
-                        // Only decrement if the value is unchanged since we started looking for work
-                        // This prevents multiple threads decrementing based on overlapping scans.
-                        //
-                        // When we decrement from active, the producer may have inserted a queue item during our scan
-                        // therefore we cannot transition to empty
-                        //
-                        // When we decrement from Maybe Inactive, if the producer inserted a queue item during our scan,
-                        // the producer must write Active.  We may transition to empty briefly if we beat the
-                        // producer's write, but the producer will then overwrite us before waking threads.
-                        // So effectively we cannot mark the queue empty when an item is in the queue.
-                        Interlocked.CompareExchange(ref WorkStealingQueueList.wsqActive, wsqActiveObserved - 1, wsqActiveObserved);
                     }
                 }
+                else
+                {
+                    // We only need to look at the global queue since WorkStealingQueueList is inactive
+                    workItems.TryDequeue(out callback);
+                }
             }
-            else
+            finally
             {
-                // We only need to look at the global queue since WorkStealingQueueList is inactive
-                workItems.TryDequeue(out callback);
+                // Update count of unclaimed work items
+                // Placed in finally block here for rare case of a thread abort exception during Dequeue
+                if (callback != null)
+                    ItemDequeued();
```

Comment:
I placed this here based on a comment about thread abort exceptions in `Dispatch()`.  If it is not really needed, it might be prettier in `Dispatch()`